src/vow.re

```diff
@@ -1,102 +1,100 @@
 module Vow = {
   type handled;
   type unhandled;
-  type t 'a 'status = {promise: Js.Promise.t 'a};
-  let return x => {promise: Js.Promise.resolve x};
-  let flatMap transform vow => {
-    promise: Js.Promise.then_ (fun x => (transform x).promise) vow.promise
-  };
-  let flatMapUnhandled transform vow => {
-    promise: Js.Promise.then_ (fun x => (transform x).promise) vow.promise
-  };
-  let map transform vow => flatMap (fun x => return (transform x)) vow;
-  let mapUnhandled transform vow => flatMapUnhandled (fun x => return (transform x)) vow;
-  let sideEffect handler vow => {
-    let _ = Js.Promise.then_ (fun x => Js.Promise.resolve @@ handler x) vow.promise;
+  type inner('a) = {inner: 'a};
+  type t('a, 'status) = Js.Promise.t(inner('a));
+  let innerReturn = (v) => {inner: v};
+  let return: 'a => t('a, 'status) = (x) => Js.Promise.resolve(innerReturn(x));
+  let flatMap: ('a => t('b, 'status), t('a, handled)) => t('b, 'status) =
+    (transform, vow) => Js.Promise.then_((x) => transform(x.inner), vow);
+  let flatMapUnhandled: ('a => t('b, 'status), t('a, unhandled)) => t('b, unhandled) =
+    (transform, vow) => Js.Promise.then_(({inner}) => transform(inner), vow);
+  let map = (transform, vow) => flatMap((x) => return(transform(x)), vow);
+  let mapUnhandled: ('a => 'b, t('a, unhandled)) => t('b, unhandled) =
+    (transform, vow) => flatMapUnhandled((x) => return(transform(x)), vow);
+  let sideEffect = (handler, vow) => {
+    let _ = Js.Promise.then_((x) => Js.Promise.resolve @@ handler(x.inner), vow);
     ()
   };
-  let onError handler vow => {
-    promise: Js.Promise.catch (fun _ => (handler ()).promise) vow.promise
-  };
-  let wrap promise => {promise: promise};
-  let unsafeWrap promise => {promise: promise};
-  let unwrap {promise} => promise;
+  let onError = (handler, vow) => Js.Promise.catch((_) => handler(), vow);
+  let wrap = (promise) => Js.Promise.then_((res) => Js.Promise.resolve(innerReturn(res)), promise);
+  let unsafeWrap = (promise) =>
+    Js.Promise.then_((res) => Js.Promise.resolve(innerReturn(res)), promise);
+  let unwrap = (promise) => Js.Promise.then_((x) => Js.Promise.resolve(x.inner), promise);
```

Comment:
I'd simply return `x` here and have the returned type as `inner`. It'd be a breaking change though.

Reply:
The problem with that is that inner('a) is a hidden by the interface, so we'd have to expose the implementation in order to return `Js.Promise.t(inner('a))`
I'm happy to do so, but not sure of the advantage.

Reply:
the advantage would be that we just return identity so that `unwrap` is simply noop.